src/lib.rs

```diff
@@ -131,26 +131,70 @@ pub fn keccakf(a: &mut [u64; PLEN]) {
     }
 }
 
-fn setout(src: &[u8], dst: &mut [u8], len: usize) {
-    dst[..len].copy_from_slice(&src[..len]);
-}
+/// Total number of lanes.
+const PLEN: usize = 25;
+
+#[derive(Default, Clone)]
+struct Buffer([u64; PLEN]);
+
+impl Buffer {
+    fn inner(&mut self) -> &mut [u64; PLEN] {
+        &mut self.0
+    }
+
+    #[cfg(target_endian = "little")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
+        let buffer: &mut [u8; PLEN * 8] = unsafe { ::core::mem::transmute(&mut self.0) };
+        f(&mut buffer[offset..][..len]);
+    }
 
-fn xorin(dst: &mut [u8], src: &[u8]) {
-    assert!(dst.len() <= src.len());
-    let len = dst.len();
-    let mut dst_ptr = dst.as_mut_ptr();
-    let mut src_ptr = src.as_ptr();
-    for _ in 0..len {
-        unsafe {
-            *dst_ptr ^= *src_ptr;
-            src_ptr = src_ptr.offset(1);
-            dst_ptr = dst_ptr.offset(1);
+    #[cfg(target_endian = "big")]
+    #[inline]
+    fn execute<F: FnOnce(&mut [u8])>(&mut self, offset: usize, len: usize, f: F) {
+        fn swap_endianess(buffer: &mut [u64]) {
+            for item in buffer {
+                let i: &mut [u8; 8] = unsafe { ::core::mem::transmute(item) };
```

Comment:
Can't we use `item.to_be()`? Or is this code here in place to avoid copying `u64`?

Reply:
`to_be` is no-op on big endian platforms (and this function is executed on only on big-endian platform), but I guess I could also use `swap_bytes`. I don't think copying matters a lot here. I'll benchmark it